proposals/lip-0051.md

```diff
@@ -503,243 +644,1891 @@ crossChainTransferMessageParams = {
+
+```python
+def transferCrossChainInternal(
+    senderAddress: Address,
+    tokenID: TokenID,
+    amount: uint64,
+    receivingChainID: ChainID,
+    recipientAddress: Address,
+    messageFee: uint64,
+    data: str
+) -> None:
+
+    # Debit the user's account
+    availableBalance(senderAddress, tokenID) -= amount
+    
+    # Escrow has to be updated only if the token is native to the chain.
+    if getChainID(tokenID) == OWN_CHAIN_ID:       
+        escrowAmount(receivingChainID, tokenID) += amount
+
+    emitEvent(
+        module=MODULE_NAME_TOKEN,
+        name=EVENT_NAME_TRANSFER_CROSS_CHAIN,
+        data={
+            "senderAddress": senderAddress,
+            "recipientAddress": recipientAddress,
+            "tokenID": tokenID,
+            "amount": amount,
+            "receivingChainID": receivingChainID,
+            "result": RESULT_SUCCESSFUL
+        },
+        topics=[senderAddress, recipientAddress, receivingChainID]
+    )
+    
+    # call interop. module to create CCM
+    Interoperability.send(
+        sendingAddress=senderAddress,
+        module=MODULE_NAME_TOKEN,
+        crossChainCommand=CROSS_CHAIN_COMMAND_NAME_TRANSFER,
+        receivingChainID=receivingChainID,
+        fee=messageFee,
+        status=CCM_STATUS_OK,
+        params={  
+                "tokenID": tokenID,
+                "amount": amount,
+                "senderAddress": senderAddress,
+                "recipientAddress": recipientAddress,
+                "data": data
+            }
+    )
+```
+
+
+### Protocol Logic for Other Modules
+
+The Token module provides the following methods to modify the token state. Any other modules should use those to modify the token state. The token state should never be modified from outside the module without using one of the proposed functions as this could result in unexpected behavior and could cause an improper state transition.
+
+#### isNativeToken
+    
+```python
+def isNativeToken(tokenID: TokenID) -> bool:
+    return getChainID(tokenID) == OWN_CHAIN_ID
+```
+
+#### userSubstoreExists
+
+```python
+def userSubstoreExists(address: Address, tokenID: TokenID) -> bool:
+    if userStore(address, tokenID) exist:
+        return True
+    else:
+        return False
+```
+
+
+#### getAvailableBalance
+
+```python
+def getAvailableBalance(address: Address, tokenID: TokenID) -> uint64:
+    return availableBalance(address, tokenID)
+```
+
+#### getLockedAmount
+
+```python
+def getLockedAmount(address: Address, module: Module, tokenID: TokenID) -> uint64:
+    return lockedAmount(address, module, tokenID)
+```
+
+#### getEscrowedAmount
+
+```python
+def getEscrowedAmount(escrowChainID: ChainID, tokenID: TokenID) -> uint64:
+    if getChainID(tokenID) != OWN_CHAIN_ID:
+        raise Exception("Token ID is not from the current chain")
+    
+    return escrowAmount(escrowChainID, tokenID)
+```
+
+
+#### initializeToken
+
+```python
+def initializeToken() -> TokenID:
```

Comment:
I think it would be better to allow to pass the desired tokenID as input and just check if it was used before (this is easy by a single read from the supply store). This is more consistent to what we currently do with chain IDs in interoperability (users can select the desired value as long as it is free), and it also slightly simplifies the protocol. 

I don't think it is really a problem that you would have to try several values, because this is arguably solved in the UI (same reasoning as in interoperability)

Reply:
If we go for this, we would remove the `MAX_AVAILABLE_ID_REACHED` result and just add a `TOKEN_ID_ALREADY_ASSIGNED` result (or smtg like that)

Reply:
I think this is the only remaining open point now. I suggest discussing this with @MaximeGagnebin on Thursday and reach a conclusion.

Reply:
For me the main drawback is that you don't have a good way to get an available ID. This was solved before with the  `nextAvailableID` store, but we removed this one.
To be clear, I think that we should really have a non-failing token initialization, or at least a way to call it without the risk of it failing. The reason being that this API call could happen as part of something way more general and you don't want the user to have to deal with the tokenID.

In general, I think I prefer the way it is now. If we are using UI arguments, you could also say that "whatever tokenID you get assigned, the UI will give it a proper name and will display it in a meaningful way".

Reply:
In the current version (after discussing within research last week) the following solution is chosen:

- we have a function `isTokenIDAvailable(tokenID)` to tell is the specified ID can be used
- `initializeToken(tokenID)` fails in case the `tokenID` is already used.
- Therefore `tokenID` is chosen by developers and is not consecutive.
- `getNextAvailableTokenID()` gives the ID after the maxID used until now. Therefore in case someone uses `tokenID` equal to the maximum uint32 value, then this function returns that there are no available tokenIDs..

@ricott1 @shuse2 @MaximeGagnebin Let me know if you think there is something extra to discuss here, or you think it can be merged like this..

Reply:
this looks fine to me

Reply:
LGTM

Reply:
LGTM